solr/solrj/src/java/org/apache/solr/common/util/CommonTestInjection.java

```diff
@@ -75,4 +78,49 @@ public static boolean injectDelay() {
     }
     return true;
   }
+
+  /**
+   * This should ONLY be set from unit test cases.
+   *
+   * Block at certain code execution point (defined by the key at which injectWaitForCondition was inserted)
+   * until the provided condition is fulfilled
+   * 
+   * @see CommonTestInjection#injectWaitForCondition(String)
+   * @param key
+   * @param condition
+   */
+  public static void setWaitForCondition(String key, Condition condition) {
+    synchronized (waitForConditions) {
```

Comment:
Could we use ConcurrentHashMap here instead of externally synchronizing on the `waitForConditions` map?

Reply:
I used this as i wanted to reduce overhead of `waitForConditions.get` as much as possible as `setWaitForCondition` is going to get called only from unit testing, though now i think even that idea is broken, cause write and read can still happen concurrently so things could still break 😓 

Yes I think it makes sense to just use concurrent hash map now, especially that we are going to guard the check with assert 😊